### PR TITLE
Minor improvements in Python API and testing

### PR DESCRIPTION
**Pull request by paulromano**
This pull request makes an assortment of small improvements and fixes:

- Make the default `TestingHarness._cleanup()` method take care of tallies.xml files. This makes it such that classes that are inheriting TestingHarness usually don't have to do anything special for cleanup.
- Store the full XML in inputs_true.dat for tests. It doesn't take up that much space, and it will make it much easier to see what changed in the inputs we are messing around with tests. Right now, when a test that uses the Python API changes, you get a new (meaningless) hash and if you want to know what actually changed, you have to figure it out yourself.
- Make sure a density has been set for a material before exporting to XML.
- Simplify representation of complex regions. For example, if you have three surface half-spaces a, b, and c and you assign a region `cell.region = a & b & c`, you'll currently end up getting two intersections, i.e. `((a & b) & c)`. Now the logic will recognize this as an intersection of three half-spaces so that the XML exported doesn't require parentheses.
- Get rid of WRITTEN_IDS module variable 
- Change how statepoint and sourcepoint options are specified in the openmc.Settings() object. There are now simply `Settings.sourcepoint` and `Settings.statepoint` attributes that expect dictionaries. I've also removed the "interval" option from the Fortran side -- much simpler to just use `range()` on the Python side.
- Make Shannon entropy mesh and UFS mesh specified using a `openmc.Mesh` rather than multiple attributes in `openmc.Settings`.
- Assorted documentation updates.

### CODE REVIEW

File: openmc/cell.py [hunk 1]
```diff
@@ -557,10 +561,9 @@ def create_xml_subelement(self, xml_element):
             # thus far.
             def create_surface_elements(node, element):
                 if isinstance(node, Halfspace):
-                    path = './surface[@id=\'{0}\']'.format(node.surface.id)
+                    path = "./surface[@id='{}']".format(node.surface.id)
```
Comment (wbinventor) [inline openmc/cell.py]: Interesting - just curious what difference the double quotes makes?

File: openmc/geometry.py [hunk 2]
```diff
@@ -73,19 +73,16 @@ def export_to_xml(self, path='geometry.xml'):
             Path to file to write. Defaults to 'geometry.xml'.
 
         """
-        # Clear OpenMC written IDs used to optimize XML generation
-        openmc.universe.WRITTEN_IDS = {}
```
Comment (wbinventor) [inline openmc/geometry.py]: If I recall correctly, @nhorelik introduced this to greatly accelerate the exportation of large geometries. Have you tried exporting BEAVRS with this change? Is it any slower?

File: openmc/cell.py (hunk 1 above)
Comment (paulromano) [inline openmc/cell.py]: Just allows the single quotes within to not be escaped

File: openmc/geometry.py (hunk 2 above)
Comment (paulromano) [inline openmc/geometry.py]: On my laptop -- exporting BEAVRS using develop branch is 7.2 sec and with this one is 9.0 sec, so it is measurably slower but still same order of magnitude.

File: openmc/cell.py (hunk 1 above)
Comment (wbinventor) [inline openmc/cell.py]: Oh interesting, I actually never realized that.

**Comment by wbinventor** [comment]
This PR looks good to me.

**Pull request merged**